# Sample Data Wrangling

For the module nine project you are required to produce coding samples for how you wrangled, ie cleaned, your dataset as one piece of your project submission. This Notebook is provided with some sample data wrangling from module two. You may copy examples from the code blocks herein to include in your final submission.

<div class="alert alert-block alert-info">
    <span style="color:black"><b><i>Activity Code Submission:</i></b> You will complete your assignment code in the <b>Module_Nine_Project.ipynb</b> Notebook file from the filetree on the left that is pre-templated for your modification in accordance with the assignment instructions.</span>
</div>

### Data Source Information
The sourcing information for the data sets can be found in the `README.md` file within the data folder.

To use this sample script for your selected data set, you can copy the desired code blocks to the Module_Nine_Project Notebook that you will use for your project submission, or replace the sample data path below keeping in mind you will need to copy the items to the required Project Notebook for inclusion in your project submission.

<div class="alert alert-block alert-success">
    <span style="color:black"><b><i>To Begin:</i></b> replace the <mark>data/sample_use_only/titanic_raw.csv</mark> sample data path with the path to your selected data set.</span>
</div>

In [1]:
reference_file = "data/sample_use_only/titanic_raw.csv"

## Installing Required Packages
If you have additional packages that you would like to use or install, add them to the **requirements.txt** file then run the next block.

In [2]:
!pip install --upgrade pip

In [3]:
!pip install -r sample_requirements.txt

## Import the Required Packages
If you have additional packages that you would like to import, add them here.

In [4]:
import pandas as pd
import numpy as np

## Loading the Data
After you have selected your data set that wish to wrangle, replace the reference file path in the code block at the top of this Notebook, then run the following block.

In [5]:
raw_data = pd.read_csv(reference_file)
raw_data.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


# Cleaning the Data
The sections that follow have some pre-templated steps to examine and clean data. The columns or specified attributes will need to be replaced for your respective dataset that you are cleaning. You may also need to add different cleaning options from your reading as this provides an example starting point to work with your selected dataset.

***Remember to add comments as you go.***

## Columns

<div class="alert alert-block alert-danger">
    <span style="color:black"><b><i>Example Data:</i></b> Keep in mind that the column values are from the sample data and will not represent your dataset until these blocks are run.</span>
</div>

In [6]:
raw_data.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

### Renaming Columns
If any of the columns in your dataset need renamed, you can use the following examples to update your column names.

<div class="alert alert-block alert-danger">
    <span style="color:black"><b><i>Example Data:</i></b> Keep in mind that the column values selected below are from the sample data and will need replaced before running the block with your dataset.</span>
</div>

In the example below we want to rename the `PassengerId` column `ID` and the `Pclass` column to say `Class` as we are considering merging this data with another dataset later which will use these columns as shared keys. For this task, we use the `rename()` method and pass in a dictionary mapping the column names to their new names. We can pass `inplace=True` to change our original dataframe instead of getting a new one back, or `inplace=False` to create a new dataframe to compare against our original. In the example, we are creating a new dataframe that we will use for all follow-on cleaning actions.

In [7]:
clean_data = raw_data.rename(
    columns={
        'PassengerId': 'ID',
        'Pclass': 'Class'
    }, inplace=False
)
clean_data.head()

,ID,Survived,Class,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


## Working with Missing Values
Lets take a look at the output of `info()` method to get a quick view of additional information about the data set which can help pinpoint missing values and wrong data types.


<div class="alert alert-block alert-danger">
    <span style="color:black"><b><i>Example Data:</i></b> Keep in mind that the output below are from the sample data and will need replaced before running the block with your dataset.</span>
</div>

In the example dataset, we have an index range of 891 entries or rows available in the sample dataset. Using that foundational data, we can examine the null counts to identify which columns may have null, missing or NaN values. In the example dataset, the following columns indicate potential issues in the data: Age, Cabin, and Embarked.

In [8]:
clean_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   ID        891 non-null    int64  
 1   Survived  891 non-null    int64  
 2   Class     891 non-null    int64  
 3   Name      891 non-null    object 
 4   Sex       891 non-null    object 
 5   Age       714 non-null    float64
 6   SibSp     891 non-null    int64  
 7   Parch     891 non-null    int64  
 8   Ticket    891 non-null    object 
 9   Fare      891 non-null    float64
 10  Cabin     204 non-null    object 
 11  Embarked  889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### isnull()
We can also use the `isnull()` method to return a dataframe of boolean values where **True** represents missing data.

In [9]:
clean_data.isnull()

,ID,Survived,Class,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,False,False,False,False,False,False,False,False,False,False,True,False
1,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,True,False
3,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...
886,False,False,False,False,False,False,False,False,False,False,True,False
887,False,False,False,False,False,False,False,False,False,False,False,False
888,False,False,False,False,False,True,False,False,False,False,True,False
889,False,False,False,False,False,False,False,False,False,False,False,False


For the example dataset, since we already saw from the `info()` method that only three columns were impacted by data issues, the output of the `isnull()` method may not be as helpful so we can limit the output for a given column to check for nulls or use the output for other cleaning functions as shown in the example below.

For your selected dataset, you can replace the Column name `Age` with your selected column that you would like to assess expanding on the dataset for additional afflicted columns for cleaning.

In [10]:
bool_series = pd.isnull(clean_data["Age"])
missing_age_data = clean_data[bool_series]
missing_age_data

,ID,Survived,Class,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
17,18,1,2,"Williams, Mr. Charles Eugene",male,NaN,0,0,244373,13.0000,NaN,S
19,20,1,3,"Masselmani, Mrs. Fatima",female,NaN,0,0,2649,7.2250,NaN,C
26,27,0,3,"Emir, Mr. Farred Chehab",male,NaN,0,0,2631,7.2250,NaN,C
28,29,1,3,"O'Dwyer, Miss. Ellen ""Nellie""",female,NaN,0,0,330959,7.8792,NaN,Q
...,...,...,...,...,...,...,...,...,...,...,...,...
859,860,0,3,"Razi, Mr. Raihed",male,NaN,0,0,2629,7.2292,NaN,C
863,864,0,3,"Sage, Miss. Dorothy Edith ""Dolly""",female,NaN,8,2,CA. 2343,69.5500,NaN,S
868,869,0,3,"van Melkebeke, Mr. Philemon",male,NaN,0,0,345777,9.5000,NaN,S
878,879,0,3,"Laleff, Mr. Kristo",male,NaN,0,0,349217,7.8958,NaN,S


### Using Numpy to Find Missing Data using a Dictionary Comprehension
Rather than do this for each column, we can write a function that will use a [dictionary comprehension](https://www.python.org/dev/peps/pep-0274/) to check all the columns for us:

In [11]:
def get_nan_count(clean_data):
    return {
        col: clean_data[clean_data[col].isin([np.nan])].shape[0] for col in clean_data.columns
    }

get_nan_count(clean_data)

{'ID': 0,
 'Survived': 0,
 'Class': 0,
 'Name': 0,
 'Sex': 0,
 'Age': 177,
 'SibSp': 0,
 'Parch': 0,
 'Ticket': 0,
 'Fare': 0,
 'Cabin': 687,
 'Embarked': 2}

###  Filling Missing Values Using fillna(), replace() and interpolate()
From the example output, there were 177 rows of data that were impacted by missing data from the sample dataset. For your selected dataset you should evaluate your columns you identified with missing or null values and expand to the afflicted columns to clean your respective dataset.

#### Filling Missing Values with a Specific Value Using fillna()
Here in the example, we will fill in the missing `Age` values with an age of `0` based on dataset knowledge, you will make your own decisions for how you handle your missing data and justify your decisions as part of your activity reflection.

In [12]:
clean_data["Age"].fillna('0', inplace = True )
clean_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   ID        891 non-null    int64  
 1   Survived  891 non-null    int64  
 2   Class     891 non-null    int64  
 3   Name      891 non-null    object 
 4   Sex       891 non-null    object 
 5   Age       891 non-null    object 
 6   SibSp     891 non-null    int64  
 7   Parch     891 non-null    int64  
 8   Ticket    891 non-null    object 
 9   Fare      891 non-null    float64
 10  Cabin     204 non-null    object 
 11  Embarked  889 non-null    object 
dtypes: float64(1), int64(5), object(6)
memory usage: 83.7+ KB


If you do not find the info() method helpful to find the missing data, you can use the previously created get_nan_count() we created to view the missing data by column. In the sample dataset you will see there are still two columns with null values to be addressed.

In [13]:
get_nan_count(clean_data)

{'ID': 0,
 'Survived': 0,
 'Class': 0,
 'Name': 0,
 'Sex': 0,
 'Age': 0,
 'SibSp': 0,
 'Parch': 0,
 'Ticket': 0,
 'Fare': 0,
 'Cabin': 687,
 'Embarked': 2}

<div class="alert alert-block alert-warning">
    <span style="color:black"><b><i>Example Data:</i></b> Your selected dataset is likely to have several columns where the data is missing, create additional code blocks as necessary to clean your data according to the assignment requirements. Here we will clean the columns with different methods to show you different use cases, you will modify the code herein to the methods you feel are appropriate for your selected data set.
</div>

#### Replacing Missing Values Using replace()
Here in the example, we will replace the missing `Embarked` values with an age of `U` to represent **Unknown** based on dataset knowledge, you will make your own decisions for how you handle your missing data and justify your decisions as part of your activity reflection.

In [14]:
clean_data["Embarked"].replace(to_replace=np.nan, value='U', inplace = True )
clean_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   ID        891 non-null    int64  
 1   Survived  891 non-null    int64  
 2   Class     891 non-null    int64  
 3   Name      891 non-null    object 
 4   Sex       891 non-null    object 
 5   Age       891 non-null    object 
 6   SibSp     891 non-null    int64  
 7   Parch     891 non-null    int64  
 8   Ticket    891 non-null    object 
 9   Fare      891 non-null    float64
 10  Cabin     204 non-null    object 
 11  Embarked  891 non-null    object 
dtypes: float64(1), int64(5), object(6)
memory usage: 83.7+ KB


#### Dropping Missing Values in Pandas Using dropna()
Here in the example, we will drop the column `Cabin` based on dataset knowledge, you will make your own decisions for how you handle your missing data and justify your decisions as part of your activity reflection. Keep in mind that if you use dropna() with an axis of 1, where there are nulls or missing values in more than one column, it may drop all columns that have null or missing values.

In [15]:
clean_data.dropna(axis=1, inplace = True )
clean_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   ID        891 non-null    int64  
 1   Survived  891 non-null    int64  
 2   Class     891 non-null    int64  
 3   Name      891 non-null    object 
 4   Sex       891 non-null    object 
 5   Age       891 non-null    object 
 6   SibSp     891 non-null    int64  
 7   Parch     891 non-null    int64  
 8   Ticket    891 non-null    object 
 9   Fare      891 non-null    float64
 10  Embarked  891 non-null    object 
dtypes: float64(1), int64(5), object(5)
memory usage: 76.7+ KB


### Duplicated Records
Some datasets may include duplicated records. 

<div class="alert alert-block alert-danger">
    <span style="color:black"><b><i>Example Data:</i></b> Keep in mind that the data types are from the sample data and will not represent your dataset until these blocks are run.</span>
</div>

The sample dataset does not have any duplicated records, though some code examples will be provided here for your use against your selected dataset, update the actions you would like to take against your selected dataset to clean it to your preference.

In [16]:
clean_data.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Length: 891, dtype: bool

## Data Types
<div class="alert alert-block alert-danger">
    <span style="color:black"><b><i>Example Data:</i></b> Keep in mind that the data types are from the sample data and will not represent your dataset until these blocks are run.</span>
</div>

You can use either the .dtypes property to view the data types for each column in the data frame, or the .info() method which will provide additional insights into the columns.

In [17]:
clean_data.dtypes

ID            int64
Survived      int64
Class         int64
Name         object
Sex          object
Age          object
SibSp         int64
Parch         int64
Ticket       object
Fare        float64
Embarked     object
dtype: object

### Data Type Conversion
If any of the data types in your dataset need to be modified or redefined, you can use the following block to update your column data types.

<div class="alert alert-block alert-danger">
    <span style="color:black"><b><i>Example Data:</i></b> Keep in mind that the columns and data types selected below are from the sample data and will need replaced before running the block with your dataset.</span>
</div>

In the example below we want to modify the data values and type for the `Survived` column to represent a `bool`.

If you do not have any type conversions required for your data set, feel free to comment out the existing example and add a comment as applicable in the code block below:

In [18]:
clean_data = clean_data.assign(
    Survived=lambda x: x.Survived.astype('bool')
)
clean_data.dtypes

ID            int64
Survived       bool
Class         int64
Name         object
Sex          object
Age          object
SibSp         int64
Parch         int64
Ticket       object
Fare        float64
Embarked     object
dtype: object

We will take a quick look at the output of the describe method.

In [19]:
clean_data.describe()

,ID,Class,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,891.000000,891.000000
mean,446.000000,2.308642,0.523008,0.381594,32.204208
std,257.353842,0.836071,1.102743,0.806057,49.693429
min,1.000000,1.000000,0.000000,0.000000,0.000000
25%,223.500000,2.000000,0.000000,0.000000,7.910400
50%,446.000000,3.000000,0.000000,0.000000,14.454200
75%,668.500000,3.000000,1.000000,0.000000,31.000000
max,891.000000,3.000000,8.000000,6.000000,512.329200


We will now take a look at the top five records in our dataframe before continue the cleaning examples.

In [20]:
clean_data.head()

,ID,Survived,Class,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
0,1,False,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S
1,2,True,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C
2,3,True,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S
3,4,True,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S
4,5,False,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S


### Categoricals
When choosing which columns to set as categories, you will want to evaluate how many different values are available for the column.
<div class="alert alert-block alert-danger">
    <span style="color:black"><b><i>Example Data:</i></b> Keep in mind that the output are from the sample data and will not represent your dataset until these blocks are run.</span>
</div>

Here in the example dataset, there are three values for the `Class` column, there are two values for the `Sex` column, and three values for the `Embarked` column, so we will convert those to categories for furture analysis. All other columns aside from `Survived` column that we converted to a bool, have a variety of values which cannot be easily placed into categories without first creating bins to use as categoricals which will we learn more about in the coming modules.

In [21]:
clean_data_with_categories = clean_data.assign(
    Class=clean_data.Class.astype('category'),
    Sex=clean_data.Sex.astype('category'),
    Embarked=clean_data.Embarked.astype('category')
)
clean_data_with_categories.dtypes

ID             int64
Survived        bool
Class       category
Name          object
Sex         category
Age           object
SibSp          int64
Parch          int64
Ticket        object
Fare         float64
Embarked    category
dtype: object

In [22]:
clean_data_with_categories.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   ID        891 non-null    int64   
 1   Survived  891 non-null    bool    
 2   Class     891 non-null    category
 3   Name      891 non-null    object  
 4   Sex       891 non-null    category
 5   Age       891 non-null    object  
 6   SibSp     891 non-null    int64   
 7   Parch     891 non-null    int64   
 8   Ticket    891 non-null    object  
 9   Fare      891 non-null    float64 
 10  Embarked  891 non-null    category
dtypes: bool(1), category(3), float64(1), int64(3), object(3)
memory usage: 52.8+ KB


In [23]:
clean_data_with_categories.describe()

,ID,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,891.000000
mean,446.000000,0.523008,0.381594,32.204208
std,257.353842,1.102743,0.806057,49.693429
min,1.000000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,0.000000,31.000000
max,891.000000,8.000000,6.000000,512.329200


In [24]:
clean_data_with_categories.describe(include='category')

,Class,Sex,Embarked
count,891,891,891
unique,3,2,4
top,3,male,S
freq,491,577,644


## Tabular Format
<div class="alert alert-block alert-danger">
    <span style="color:black"><b><i>Example Data:</i></b> Keep in mind that the output are from the sample data and will not represent your dataset until these blocks are run. Validate your table represents your dataset before activity submission.</span>
</div>

In [25]:
len(clean_data_with_categories)

891

In [26]:
pd.set_option('display.max_rows', None)
clean_data_with_categories

,ID,Survived,Class,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
0,1,False,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S
1,2,True,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C
2,3,True,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S
3,4,True,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S
4,5,False,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S
5,6,False,3,"Moran, Mr. James",male,0,0,0,330877,8.4583,Q
6,7,False,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,S
7,8,False,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,S
8,9,True,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,S
9,10,True,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,C
